ClientPool/ClientPool/EventStoreClientPool.cs

```diff
@@ -0,0 +1,135 @@
+        RETRY_READ:
+
+            try {
+                _log.LogTrace("Reading from: Node-{@ReadNode} - {@NodeUri}", node.ReadNodeIndex, node.ServerUri);
+                var slice = node.Client.ReadStreamAsync(direction, streamName, revision, maxCount, resolveLinkTos, deadline, userCredentials, token);
+                events.AddRange(await slice.ToListAsync(token));
+            } catch (Exception exc) {
+                numberOfRetries++;
+                if (numberOfRetries == _readers.Count) throw;
+                _log.LogWarning(exc, "Read failed for current node.  Retrying on next node.");
+                node = NextNode();
+                goto RETRY_READ;
+            }
+
+        RETRY_LEADER_READ:
+            try {
+                if (evt.Event != null) {
+                    var slice = _leader.Client.ReadStreamAsync(direction, streamName, evt.OriginalEventNumber, maxCount, resolveLinkTos, deadline, userCredentials, token);
+                    events.AddRange(await slice.ToListAsync());
+                }
+            } catch (NullReferenceException) {
+                // squelch.
+            } catch (NotLeaderException exc) {
+                if (retriedLeaderRead) throw;
+                retriedLeaderRead = true;
+                BuildLeaderConnection(exc);
+                goto RETRY_LEADER_READ;
+            }
+
+            return events;
+        }
+        private ReadNodeInformation NextNode() {
+            _readSemaphor.Wait();
+
+            try {
+                if (_readersEnumerator.Current == null) {
+                    _readersEnumerator = _readers.GetEnumerator();
+                    _readersEnumerator.MoveNext();
+                    return _readersEnumerator.Current;
+                }
+
+                if (_readersEnumerator.MoveNext()) return _readersEnumerator.Current;
+
+                _readersEnumerator = _readers.GetEnumerator();
+                _readersEnumerator.MoveNext();
+                return _readersEnumerator.Current;
+            } finally {
+                _readSemaphor.Release();
+            }
+        }
+
+        private void BuildLeaderConnection(NotLeaderException exc) {
+            _log.LogWarning(exc, "Re-configuring leader to resolved node.");
+            var llNode = _readers.First(x => x.ServerUri.Host == exc.LeaderEndpoint.Host && x.ServerUri.Port == exc.LeaderEndpoint.Port);
+            if (llNode == null) throw exc;
+
+            BuildLeaderConnection(llNode.ServerUri);
+        }
+
+        private void BuildLeaderConnection(Uri leaderUri) {
+            _log.LogInformation("Setting leader connection to {@leaderUri}.", leaderUri);
+            var grpcClientSettings = EventStoreClientSettings.Create($"{_options.LeaderUri}");
+            grpcClientSettings.DefaultCredentials = _options.DefaultCredentials;
+            grpcClientSettings.ConnectivitySettings.NodePreference = Client.NodePreference.Leader;
```

Comment:
```suggestion
            grpcClientSettings.ConnectivitySettings.NodePreference = NodePreference.Leader;
```
Otherwise it does not build